sys-fs/cryfs - new package/proxy maintenance request

automorphism88 opened the pull request:
I previously opened a bug on Bugzilla for this package (https://bugs.gentoo.org/show_bug.cgi?id=612688) and was told on IRC that I should submit a PR on Github.

I am willing to volunteer as a proxy maintainer for this package. It is the first one I have submitted, so any feedback would be most welcome.

automorphism88 commented:
The issue identified by @gentoo-repo-qa-bot should be fixed by this commit: https://github.com/gentoo/gentoo/pull/4624/commits/c2ff30a3c9e0bab8732540f1ddb02d8076787a86

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+# Copyright 1999-2017 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=5
```
Go for 6, please.

automorphism88 replied:
Done, but just wondering, why? I looked at the list of EAPI 6 changes and there didn't seem to be anything that was required.

mgorny replied:
We always want to use the newest EAPI available. This may expose newer API in eclasses, get rid of some deprecated bits but most importantly it gets us one step closer to cleaning up old EAPIs. While 5-6 indeed isn't that significant, a few years from now we'd like to avoid having to put more effort into cleaning up packages that were already added with old EAPI.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=5
+PYTHON_COMPAT=( python{2_7,3_{4,5,6}} )
```
Have you tested every single one of them wrt https://wiki.gentoo.org/wiki/Project:Python/Adding_and_Removing_Python_implementations#Testing_support_for_the_implementation?

automorphism88 replied:
Yes, I did. Note that Python is only used at build time. But I was able to build it successfully with all four.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+HOMEPAGE="https://www.cryfs.org/"
+
+SLOT=0
+IUSE="-update-check -test -debug"
```
`-flag` form is undesired (and possibly invalid). The flags are off by default, so just list them without prefixes. Sort lexically.

automorphism88 replied:
Fixed.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+IUSE="-update-check -test -debug"
+
+LICENSE="LGPL-3 BSD-2 MIT
+	test? ( BSD )"
```
LICENSE only applies to installed files, so you can skip test code there.

automorphism88 replied:
Fixed.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+	test? ( BSD )"
+# cryfs - LGPL-3
+# scrypt - BSD-2
+# spdlog - MIT
```
Does that mean it bundles those libraries?

automorphism88 replied:
Yes, it unfortunately bundles scrypt and spdlog. I've requested upstream that CMake options be added to use the system versions instead. If that isn't done then I'll look at doing it myself in the ebuild. It should be easy to replace spdlog with the system version, but to use the system scrypt, the app-crypt/scrypt ebuild would need to be modified to install some header/source files, as right now it only installs the binary /usr/bin/scrypt.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+	>=dev-libs/crypto++-5.6.3:=
+	net-misc/curl:=
+	>=sys-fs/fuse-2.8.6:=
+	dev-libs/openssl:="
```
If it needs openssl headers, you should bind it to slot 0 (i.e. `:0=`).

automorphism88 replied:
Fixed.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+	dev-libs/openssl:="
+DEPEND="${RDEPEND}
+	${PYTHON_DEPS}
+	|| ( >=sys-devel/gcc-4.8 >=sys-devel/clang-3.7 )"
```
Compiler dependencies are not meaningful. Plus, we no longer support compilers that old.

automorphism88 replied:
Fixed.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+DEPEND="${RDEPEND}
+	${PYTHON_DEPS}
+	|| ( >=sys-devel/gcc-4.8 >=sys-devel/clang-3.7 )"
+CMAKE_MIN_VERSION="2.8"
```
I think this goes before inherit. However, we do not have cmake that old, so you can just omit it.

automorphism88 replied:
Removed that line.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+CMAKE_MIN_VERSION="2.8"
+
+src_configure() {
+	local mycmakeargs=("-DBoost_USE_STATIC_LIBS=off")
```
Why don't you make it one big array, and use `$(usex foo)` to switch between yes/no? Also note that for cmake, on/yes/... and off/no/... are equivalent.

automorphism88 replied:
Done.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+		mycmakeargs+=("-DBUILD_TESTING=on")
+	fi
+
+	if use debug ; then
```
Does upstream specifically play with CMAKE_BUILD_TYPE? We normally don't use `use debug` like that, especially that we disarm most of CMAKE_BUILD_TYPE-conditional logic (like altering flags).

automorphism88 replied:
Yes. Without setting the CMAKE_BUILD_TYPE it doesn't build correctly. When building it outside of portage, the CMAKE_BUILD_TYPE is set to Release by default, but it was getting set to "Gentoo" by portage when I didn't include this declaration, breaking the build.

The debug USE flag could be removed if it's not supposed to be used this way. I just figured the option provided by upstream should be made available to the user.

mgorny replied:
Does upstream do anything special wrt Debug/Release? We usually recommend `RelWithDebInfo` if there is no special case for the other options. This usually disables most of the undesired mangling.

mgorny replied:
No, that's not a very good solution. We use those only as a last resort.

FWICS, the build system is very stupid and assumes stuff ought only to be installed in `Release` mode. They do not seem to be using CMAKE_BUILD_TYPE for anything else, and we disarm CMake builtins, so just put `CMAKE_BUILD_TYPE=Release` in there and be done with it.

(this also implies that you can't use `CMAKE_BUILD_TYPE=Debug` as then none of the files would be installed)

automorphism88 replied:
Alright, I got rid of the src_install() function I'd added and just put an unconditional `CMAKE_BUILD_TYPE=Release` in.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+}
+
+src_test() {
+	TMPDIR="${T}"
```
`local`. Plus, PMS already requires the package manager to alter TMPDIR.

automorphism88 replied:
l made it local. I'm not sure why, but without declaring TMPDIR here, the sandbox causes problems.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+src_test() {
+	TMPDIR="${T}"
+	addread /dev/fuse
+	addwrite /dev/fuse
```
Are those tests safe?

automorphism88 replied:
As far as I know; there didn't seem to be any ill effects when run on my system, and the tests don't work without allowing read/write access to /dev/fuse.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+	addread /dev/fuse
+	addwrite /dev/fuse
+
+	for i in gitversion cpp-utils parallelaccessstore blockstore blobstore fspp cryfs cryfs-cli ; do
```
Doesn't upstream provide some make target to run tests?

automorphism88 replied:
Not to my knowledge. Enabling the CMake test option only causes the binaries that this for loop runs to be built. I also looked through the Makefile generated by cmake and the upstream documentation. I could be missing something though.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+		${BUILD_DIR}/test/${i}/${i}-test || die "${i}-test failed"
+	done
+
+	adddeny /dev/fuse
```
I'm not sure if that is effective, i.e. if addwrite doesn't take precedence.

automorphism88 replied:
I'm not sure either, but when I was trying to figure out how to make the tests work in the sandbox, Google led me to a discussion of another FUSE-related package with a similar issue, where it was suggested to do addread/addwrite before the tests and adddeny afterwards.

mgorny commented on sys-fs/cryfs/metadata.xml:
```diff
@@ -0,0 +1,15 @@
+  </use>
+  <longdescription>Encrypted FUSE filesystem that conceals metadata</longdescription>
+</pkgmetadata>
+    
```
Stray line.

automorphism88 replied:
Fixed

automorphism88 commented:
What's the status of this PR? I believe I've fixed all of the issues identified, except for the bundled libraries. I'm still waiting to hear back from upstream about that, but if it's the only issue preventing the ebuild from being merged then I'll go ahead and work on patching it myself, if I can get some guidance on the best way to handle the scrypt header issue (new USE flag for the existing scrypt ebuild perhaps?).

If there's anything else that needs to be changed before the ebuild is suitable to be merged, or if my explanations for the stuff I didn't change (e.g. the use of CMAKE_BUILD_TYPE) weren't satisfactory, please let me know so that I can fix it.

Also, there is currently an upstream issue that causes two of the tests run by "cpp-utils-test" to fail due to an invalid domain being given to curl. Because of this I changed the src_test function so that instead of immediately calling die if one of the test binaries fails, it uses a local variable to track failed tests and calls die at the end if one or more tests has failed, after reporting which tests failed with eerror. This way the function will attempt all the tests no matter what.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,70 @@
+src_configure() {
+	local mycmakeargs=( "-DBoost_USE_STATIC_LIBS=off"
+						"-DCRYFS_UPDATE_CHECKS=$(usex update-check)"
+						"-DBUILD_TESTING=$(usex test)" )
```
I'm not sure what you tried to achieve here with this indent but it looks horrible ;-). Please use a single tab-indent per level, e.g.:

```
    local mycmakeargs=(
        -Dfoo=off
        -Dbar=$(usex frobnicate)
    )

automorphism88 replied:
It made more sense back when there was only one element in the array :) Fixed.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,70 @@
+	local TMPDIR="${T}"
+	addread /dev/fuse
+	addwrite /dev/fuse
+	local tests_failed=""
```
Use a bash array here, i.e.:

```
local tests_failed=()
...
tests_failed+=( foo )
...
if [[ -n ${tests_failed[@]} ]]; then ...
```

automorphism88 replied:
Done.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,70 @@
+	local tests_failed=""
+
+	for i in gitversion cpp-utils parallelaccessstore blockstore blobstore fspp cryfs cryfs-cli ; do
+		${BUILD_DIR}/test/${i}/${i}-test || tests_failed+="${i} "
```
BUILD_DIR needs to be quoted.

automorphism88 replied:
Done.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,70 @@
+
+	adddeny /dev/fuse
+
+	if [ -n "$tests_failed" ] ; then
```
Always use bash [[ ]] thingies.

automorphism88 replied:
Changed it.

mgorny commented:
> Also, there is currently an upstream issue that causes two of the tests run by "cpp-utils-test" to fail due to an invalid domain being given to curl.

The ebuild (during build or testing) must not access Internet in any way. If that is the case, you need to either disable (remove) those tests, or `RESTRICT=test` entirely (assuming all tests require Internet).

automorphism88 commented:
I've added a src_prepare() function that removes the tests that require internet access using sed, and also fixed the other things. I'll push my changes to github as soon as I've confirmed the test build I'm currently running worked.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -36,10 +36,20 @@ RDEPEND=">=dev-libs/boost-1.56:=
+src_prepare() {
+	# remove tests that require internet access to comply with Gentoo policy
+	sed -i '/CurlHttpClientTest.cpp/d' test/cpp-utils/CMakeLists.txt || die
+	sed -i '/FakeHttpClientTest.cpp/d' test/cpp-utils/CMakeLists.txt || die
```
Minor nitpick but you can pass multiple `-e` expressions to sed, i.e. `sed -i -e /foo/d -e /bar/d ...`

automorphism88 replied:
Combined them

mgorny commented on sys-fs/cryfs/cryfs-9999.ebuild:
```diff
@@ -54,15 +64,15 @@ src_test() {
-	if [ -n "$tests_failed" ] ; then
+	if [[ -n ${tests_failed[@]} ]] ; then
 		eerror "The following tests failed:"
 		eerror "$tests_failed"
```
Here you also need `[@]` or `[*]`, or it will only print the first element.

automorphism88 replied:
Whoops, forgot that when I was changing the other stuff to make it use a bash array instead of a single string. I've gotten in the habit of avoiding bashisms in shell scripts where possible, hence this and the single bracket test.

mgorny commented on sys-fs/cryfs/cryfs-0.9.7.ebuild:
```diff
@@ -0,0 +1,75 @@
+		"-DBUILD_TESTING=$(usex test)"
+	)
+
+	CMAKE_BUILD_TYPE=Release
```
You can make it `local`, it's used by src_configure only.

mgorny commented:
Oops, it seems that I have merged an earlier version. I'll fix it in a minute.

automorphism88 commented:
Thanks for your help in revising this ebuild. I learned a lot in the process and hope to contribute more.
